src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -530,6 +540,99 @@ int sd_dhcp_client_set_max_attempts(sd_dhcp_client *client, uint64_t max_attempt
         return 0;
 }
 
+static void dhcp_option_hash_func(const sd_dhcp_option *id, struct siphash *state) {
+         assert(id);
+         assert(id->option);
+         assert(id->length);
+         assert(id->data);
+
+         siphash24_compress(&id->option, sizeof(id->option), state);
+         siphash24_compress(&id->length, sizeof(id->length), state);
+         siphash24_compress(id->data, id->length, state);
+}
+
+static int dhcp_option_compare_func(const sd_dhcp_option *a, const sd_dhcp_option *b) {
+         int r;
+
+         assert(!a->length || a->data);
+         assert(!b->length || b->data);
+
+         r = CMP(a->option, b->option);
+         if (r != 0)
+                 return r;
+
+         r = CMP(a->length, b->length);
+         if (r != 0)
+                 return r;
+
+         return memcmp(a->data, b->data, a->length);
+ }
+
+static sd_dhcp_option* dhcp_option_free(sd_dhcp_option *i) {
+        if (!i)
+                return NULL;
+
+        free(i->data);
+        return mfree(i);
+}
+
+DEFINE_PRIVATE_HASH_OPS(dhcp_option_hash_ops, sd_dhcp_option, dhcp_option_hash_func, dhcp_option_compare_func);
+
+_public_ int sd_dhcp_option_new(sd_dhcp_option **ret) {
+        sd_dhcp_option *p;
+
+        assert_return(ret, -EINVAL);
+
+        p = new(sd_dhcp_option, 1);
+        if (!p)
+                return -ENOMEM;
+
+        *p = (sd_dhcp_option) {
+                .n_ref = 1,
+         };
+
+        *ret = p;
+        return 0;
+}
+
+DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC(sd_dhcp_option, sd_dhcp_option, dhcp_option_free);
+
+int sd_dhcp_client_add_dhcp_option(sd_dhcp_client *client, sd_dhcp_option *v) {
+        int r;
+
+        assert_return(client, -EINVAL);
+        assert_return(v, -EINVAL);
+
+        r = set_ensure_allocated(&client->options, &dhcp_option_hash_ops);
+        if (r < 0)
+                return -ENOMEM;
+
+        r = set_put(client->options, v);
```

Comment:
I wonder if it makes sense to build the infra with hashing and so on. It would be simpler to keep them in an array. Why would we care if the caller tries to attach the same option twice? And in some cases, it might be ever useful to allow that.

Reply:
In what cases ? can you come up with use case which allows duplicate data ?

Reply:
I can't come up with a use case. There are various DHCP options out there, so there might be various ones which make some use of this. But this is just a secondary consideration. Not making the code complicated when it doesn't have to be is more important. (In particular, an array is *much* easier to debug.)

Reply:
I asked for this because the DHCP options are not duplicated and I could not find any Wireshark dump are searching which says duplicate values are allowed and valid. The library is not also capable of reading duplicate values and stops where it finds. It's not about debug but doing it correctly.